In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load in 

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the "../input/" directory.
# For example, running this (by clicking run or pressing Shift+Enter) will list the files in the input directory

import os
print(os.listdir("../input"))
import matplotlib.pyplot as plt
%matplotlib inline
# Any results you write to the current directory are saved as output.

['sample_submission.csv', 'act_test.csv', 'act_train.csv', 'people.csv']


In [2]:
act_train = pd.read_csv('../input/act_train.csv', parse_dates = ['date'])
act_test = pd.read_csv('../input/act_test.csv', parse_dates = ['date'])
people = pd.read_csv('../input/people.csv', parse_dates = ['date'])

In [3]:
people.head()

people_id  char_1      group_1   ...   char_36 char_37 char_38
0     ppl_100  type 2  group 17304   ...      True   False      36
1  ppl_100002  type 2   group 8688   ...      True   False      76
2  ppl_100003  type 2  group 33592   ...      True    True      99
3  ppl_100004  type 2  group 22593   ...      True    True      76
4  ppl_100006  type 2   group 6534   ...      True   False      84

[5 rows x 41 columns]

In [4]:
act_train.head()

people_id   activity_id       date   ...   char_9  char_10 outcome
0   ppl_100  act2_1734928 2023-08-26   ...      NaN  type 76       0
1   ppl_100  act2_2434093 2022-09-27   ...      NaN   type 1       0
2   ppl_100  act2_3404049 2022-09-27   ...      NaN   type 1       0
3   ppl_100  act2_3651215 2023-08-04   ...      NaN   type 1       0
4   ppl_100  act2_4109017 2023-08-26   ...      NaN   type 1       0

[5 rows x 15 columns]

In [5]:
act_test.head()

people_id   activity_id       date    ...       char_8   char_9    char_10
0  ppl_100004   act1_249281 2022-07-20    ...       type 7   type 4        NaN
1  ppl_100004   act2_230855 2022-07-20    ...          NaN      NaN   type 682
2   ppl_10001   act1_240724 2022-10-14    ...      type 13  type 10        NaN
3   ppl_10001    act1_83552 2022-11-27    ...       type 5   type 5        NaN
4   ppl_10001  act2_1043301 2022-10-15    ...          NaN      NaN  type 3015

[5 rows x 14 columns]

In [6]:
train_df = pd.merge(act_train, people, on = ['people_id'], how = 'left', suffixes = ('_t', '_p'))
test_df = pd.merge(act_test, people, on = ['people_id'], how = 'left', suffixes = ('_t', '_p'))

In [7]:
train_df.head()

people_id   activity_id     date_t   ...   char_36 char_37 char_38
0   ppl_100  act2_1734928 2023-08-26   ...      True   False      36
1   ppl_100  act2_2434093 2022-09-27   ...      True   False      36
2   ppl_100  act2_3404049 2022-09-27   ...      True   False      36
3   ppl_100  act2_3651215 2023-08-04   ...      True   False      36
4   ppl_100  act2_4109017 2023-08-26   ...      True   False      36

[5 rows x 55 columns]

In [8]:
test_df.head()

people_id   activity_id     date_t   ...   char_36 char_37 char_38
0  ppl_100004   act1_249281 2022-07-20   ...      True    True      76
1  ppl_100004   act2_230855 2022-07-20   ...      True    True      76
2   ppl_10001   act1_240724 2022-10-14   ...      True    True      90
3   ppl_10001    act1_83552 2022-11-27   ...      True    True      90
4   ppl_10001  act2_1043301 2022-10-15   ...      True    True      90

[5 rows x 54 columns]

In [9]:
train_df.isna().sum()/train_df.shape[0] * 100

people_id             0.000000
activity_id           0.000000
date_t                0.000000
activity_category     0.000000
char_1_t             92.826849
char_2_t             92.826849
char_3_t             92.826849
char_4_t             92.826849
char_5_t             92.826849
char_6_t             92.826849
char_7_t             92.826849
char_8_t             92.826849
char_9_t             92.826849
char_10_t             7.173151
outcome               0.000000
char_1_p              0.000000
group_1               0.000000
char_2_p              0.000000
date_p                0.000000
char_3_p              0.000000
char_4_p              0.000000
char_5_p              0.000000
char_6_p              0.000000
char_7_p              0.000000
char_8_p              0.000000
char_9_p              0.000000
char_10_p             0.000000
char_11               0.000000
char_12               0.000000
char_13               0.000000
char_14               0.000000
char_15               0.000000
char_16 

92.9% of char1 to 9 of training set are NaN values. Let's not use them in our model.

In [10]:
test_df.isna().sum()/test_df.shape[0] * 100

people_id             0.000000
activity_id           0.000000
date_t                0.000000
activity_category     0.000000
char_1_t             91.960488
char_2_t             91.960488
char_3_t             91.960488
char_4_t             91.960488
char_5_t             91.960488
char_6_t             91.960488
char_7_t             91.960488
char_8_t             91.960488
char_9_t             91.960488
char_10_t             8.039512
char_1_p              0.000000
group_1               0.000000
char_2_p              0.000000
date_p                0.000000
char_3_p              0.000000
char_4_p              0.000000
char_5_p              0.000000
char_6_p              0.000000
char_7_p              0.000000
char_8_p              0.000000
char_9_p              0.000000
char_10_p             0.000000
char_11               0.000000
char_12               0.000000
char_13               0.000000
char_14               0.000000
char_15               0.000000
char_16               0.000000
char_17 

In test set, we have to impute the values of char10

In [11]:
del act_train,act_test,people

In [12]:
train_df = train_df.assign(year=lambda train_df: train_df.date_t.dt.year,
                           month=lambda train_df: train_df.date_t.dt.month,
                           day=lambda train_df: train_df.date_t.dt.day)

In [13]:
train_df = train_df.drop('date_t', axis = 1)

In [14]:
test_df = test_df.assign(year=lambda test_df: test_df.date_t.dt.year,
                           month=lambda test_df: test_df.date_t.dt.month,
                           day=lambda test_df: test_df.date_t.dt.day)
test_df = test_df.drop('date_t', axis = 1)

In [15]:
train_df.activity_id.unique()

array(['act2_1734928', 'act2_2434093', 'act2_3404049', ..., 'act2_536973',
       'act2_688656', 'act2_715089'], dtype=object)

In [16]:
test_df.activity_id.unique()

array(['act1_249281', 'act2_230855', 'act1_240724', ..., 'act2_448830',
       'act2_450133', 'act2_847967'], dtype=object)

In [17]:
train_df.activity_id.str.startswith('act1').sum()

157615

In [18]:
train_df.activity_id.str.startswith('act2').sum()

2039676

In [19]:
assert(train_df.activity_id.nunique()==train_df.activity_id.str.startswith('act1').sum()+train_df.activity_id.str.startswith('act2').sum()), "other activities exist"

Only activity 1 and 2 exist

In [20]:
test_df.activity_id.str.startswith('act1').sum()

40092

In [21]:
test_df.activity_id.str.startswith('act2').sum()

458595

In [22]:
assert(test_df.activity_id.nunique()==test_df.activity_id.str.startswith('act1').sum()+test_df.activity_id.str.startswith('act2').sum()), "other activities exist"

In test set also, only activity 1 and 2 exist

In [23]:
train_df.activity_id.str.startswith('act1').sum()/train_df.activity_id.str.startswith('act2').sum()*100

7.727452791521791

Number of activity 1 is much lower than that of activity 2

In [24]:
train_df.activity_id = train_df.activity_id.astype(str)

In [25]:
test_df.activity_id = test_df.activity_id.astype(str)

In [26]:
new_df = train_df.copy()
new_df = new_df.assign(activity_new = new_df['activity_id'])

In [27]:
new_df = new_df.assign(activity_new2 = new_df['activity_id'])

In [28]:
new_df.activity_new = new_df.activity_new.str[3:4]

In [29]:
new_df.activity_new2 = new_df.activity_new2.str[5:]

In [30]:
new_df.activity_new.unique()

array(['2', '1'], dtype=object)

In [31]:
new_df.activity_new2.unique()

array(['1734928', '2434093', '3404049', ..., '536973', '688656', '715089'],
      dtype=object)

In [32]:
train_df = new_df
train_df = train_df.drop('activity_id', axis = 1)

In [33]:
new_df = test_df.copy()
new_df = new_df.assign(activity_new = new_df['activity_id'])
new_df = new_df.assign(activity_new2 = new_df['activity_id'])
new_df.activity_new = new_df.activity_new.str[3:4]
new_df.activity_new2 = new_df.activity_new2.str[5:]

In [34]:
test_df = new_df
test_df = test_df.drop('activity_id', axis = 1)
del new_df

In [35]:
new_df = train_df.copy()
new_df.activity_category = new_df.activity_category.str.lstrip('type ').astype(np.float)

In [36]:
new_df.loc[new_df.activity_new2=='4e+06','activity_new2'] = '4000000'

In [37]:
new_df.loc[new_df.activity_new2=='2e+06','activity_new2'] = '2000000'

In [38]:
new_df.loc[new_df.activity_new2=='2e+05','activity_new2'] = '200000'

In [39]:
new_df.loc[new_df.activity_new2=='5e+05','activity_new2'] = '500000'

In [40]:
new_df.activity_new = new_df.activity_new.replace('','-999')
new_df.activity_new = new_df.activity_new.astype('int32')
new_df.activity_new2 = new_df.activity_new2.replace('','-999')
new_df.activity_new2 = new_df.activity_new2.astype('int32')

In [41]:
new_df.loc[new_df.activity_new==-999,'activity_new'] = np.nan

In [42]:
new_df.loc[new_df.activity_new2==-999,'activity_new2'] = np.nan

In [43]:
train_df=new_df

In [44]:
new_df = test_df.copy()
new_df.activity_category = new_df.activity_category.str.lstrip('type ').astype(np.float)
new_df.loc[new_df.activity_new2=='4e+06','activity_new2'] = '4000000'
new_df.loc[new_df.activity_new2=='2e+06','activity_new2'] = '2000000'
new_df.loc[new_df.activity_new2=='2e+05','activity_new2'] = '200000'
new_df.loc[new_df.activity_new2=='5e+05','activity_new2'] = '500000'
new_df.loc[new_df.activity_new2=='3e+05','activity_new2'] = '300000'
new_df.loc[new_df.activity_new2=='9e+05','activity_new2'] = '900000'
new_df.activity_new = new_df.activity_new.replace('','-999')
new_df.activity_new = new_df.activity_new.astype('int32')
new_df.activity_new2 = new_df.activity_new2.replace('','-999')
new_df.activity_new2 = new_df.activity_new2.astype('int32')
new_df.loc[new_df.activity_new==-999,'activity_new'] = np.nan
new_df.loc[new_df.activity_new2==-999,'activity_new2'] = np.nan
test_df=new_df
del new_df

In [45]:
temp_df = train_df.copy()
col_range1 = 11
col_range2 = 10
char_columns1 = ['char_' + str(i) for i in range(1,col_range1)]
char_columns1 = [s + '_t' for s in char_columns1]
char_columns2 = ['char_' + str(i) for i in range(1,col_range2)]
char_columns2 = [s + '_p' for s in char_columns2]
char_columns3 = ['char_10_p']
char_columns3 = char_columns3+['char_' + str(i) for i in range(col_range2+1,39)]

temp_df[char_columns1] = temp_df[char_columns1].fillna('type -999')
temp_df[char_columns1] = temp_df[char_columns1].apply(lambda col: col.str.lstrip('type '))
temp_df[char_columns1] = temp_df[char_columns1].astype(np.int32)

temp_df[char_columns2] = temp_df[char_columns2].fillna('type -999').apply(lambda col: col.str.lstrip('type ')).astype(np.int32)

temp_df[char_columns3] = temp_df[char_columns3].fillna(-999)*1

In [46]:
char_columns = char_columns1+char_columns2+char_columns3
print(char_columns)

['char_1_t', 'char_2_t', 'char_3_t', 'char_4_t', 'char_5_t', 'char_6_t', 'char_7_t', 'char_8_t', 'char_9_t', 'char_10_t', 'char_1_p', 'char_2_p', 'char_3_p', 'char_4_p', 'char_5_p', 'char_6_p', 'char_7_p', 'char_8_p', 'char_9_p', 'char_10_p', 'char_11', 'char_12', 'char_13', 'char_14', 'char_15', 'char_16', 'char_17', 'char_18', 'char_19', 'char_20', 'char_21', 'char_22', 'char_23', 'char_24', 'char_25', 'char_26', 'char_27', 'char_28', 'char_29', 'char_30', 'char_31', 'char_32', 'char_33', 'char_34', 'char_35', 'char_36', 'char_37', 'char_38']


In [47]:
temp_df[char_columns] = temp_df[char_columns].replace(-999, np.NaN)

In [48]:
train_df = temp_df
del temp_df

In [49]:
temp_df = test_df.copy()
col_range1 = 11
col_range2 = 10
char_columns1 = ['char_' + str(i) for i in range(1,col_range1)]
char_columns1 = [s + '_t' for s in char_columns1]
char_columns2 = ['char_' + str(i) for i in range(1,col_range2)]
char_columns2 = [s + '_p' for s in char_columns2]
char_columns3 = ['char_10_p']
char_columns3 = char_columns3+['char_' + str(i) for i in range(col_range2+1,39)]

temp_df[char_columns1] = temp_df[char_columns1].fillna('type -999')
temp_df[char_columns1] = temp_df[char_columns1].apply(lambda col: col.str.lstrip('type '))
temp_df[char_columns1] = temp_df[char_columns1].astype(np.int32)

temp_df[char_columns2] = temp_df[char_columns2].fillna('type -999').apply(lambda col: col.str.lstrip('type ')).astype(np.int32)

temp_df[char_columns3] = temp_df[char_columns3].fillna(-999)*1

char_columns = char_columns1+char_columns2+char_columns3

temp_df[char_columns] = temp_df[char_columns].replace(-999, np.NaN)

test_df = temp_df
del temp_df

In [50]:
new_df = train_df.copy()
new_df.group_1 = new_df.group_1.str.lstrip('group ').astype(np.float)

In [51]:
train_df = new_df

In [52]:
new_df = test_df.copy()
new_df.group_1 = new_df.group_1.str.lstrip('group ').astype(np.float)
test_df = new_df
del new_df

In [53]:
test_df.group_1.unique()

array([22593., 25417., 17304., ..., 23542., 20306., 11303.])

In [54]:
train_df = train_df.assign(year_p=lambda train_df: train_df.date_p.dt.year,
                           month_p=lambda train_df: train_df.date_p.dt.month,
                           day_p=lambda train_df: train_df.date_p.dt.day)
train_df = train_df.drop('date_p', axis = 1)

In [55]:
test_df = test_df.assign(year_p=lambda test_df: test_df.date_p.dt.year,
                           month_p=lambda test_df: test_df.date_p.dt.month,
                           day_p=lambda test_df: test_df.date_p.dt.day)
test_df = test_df.drop('date_p', axis = 1)

In [56]:
train_df.head()

people_id  activity_category  char_1_t  ...    year_p  month_p  day_p
0   ppl_100                4.0       NaN  ...      2021        6     29
1   ppl_100                2.0       NaN  ...      2021        6     29
2   ppl_100                2.0       NaN  ...      2021        6     29
3   ppl_100                2.0       NaN  ...      2021        6     29
4   ppl_100                2.0       NaN  ...      2021        6     29

[5 rows x 60 columns]

We won't be able to impute activity_id. We have two options - drop the rows where it is not present or keep the rows where activity_id is null as a separate group. 

The first option is a bad one because there are multiple rows where activity_id is not present in the test set as well (see below code). If we drop in train set then, we won't be able to predict for test set.

Though the second method might be incorrect, it is the best course of action because a) we don't have domain knowledge and b) the column names are hidden. In this way, we are not biasing the results - if we impute activity id to be something, then we run the risk of biasing the results.

In [57]:
new_train_df = train_df[-(train_df.activity_new.isnull())]
new_test_df = test_df[-(test_df.activity_new.isnull())]

In [58]:
test_df[(test_df.activity_new.isnull())].count()

people_id            0
activity_category    0
char_1_t             0
char_2_t             0
char_3_t             0
char_4_t             0
char_5_t             0
char_6_t             0
char_7_t             0
char_8_t             0
char_9_t             0
char_10_t            0
char_1_p             0
group_1              0
char_2_p             0
char_3_p             0
char_4_p             0
char_5_p             0
char_6_p             0
char_7_p             0
char_8_p             0
char_9_p             0
char_10_p            0
char_11              0
char_12              0
char_13              0
char_14              0
char_15              0
char_16              0
char_17              0
char_18              0
char_19              0
char_20              0
char_21              0
char_22              0
char_23              0
char_24              0
char_25              0
char_26              0
char_27              0
char_28              0
char_29              0
char_30              0
char_31    

In [59]:
new_train_df.isna().sum()/new_train_df.shape[0] * 100

people_id             0.000000
activity_category     0.000000
char_1_t             92.826849
char_2_t             92.826849
char_3_t             92.826849
char_4_t             92.826849
char_5_t             92.826849
char_6_t             92.826849
char_7_t             92.826849
char_8_t             92.826849
char_9_t             92.826849
char_10_t             7.173151
outcome               0.000000
char_1_p              0.000000
group_1               0.000000
char_2_p              0.000000
char_3_p              0.000000
char_4_p              0.000000
char_5_p              0.000000
char_6_p              0.000000
char_7_p              0.000000
char_8_p              0.000000
char_9_p              0.000000
char_10_p             0.000000
char_11               0.000000
char_12               0.000000
char_13               0.000000
char_14               0.000000
char_15               0.000000
char_16               0.000000
char_17               0.000000
char_18               0.000000
char_19 

In [60]:
new_test_df.isna().sum()/new_test_df.shape[0] * 100

people_id             0.000000
activity_category     0.000000
char_1_t             91.960488
char_2_t             91.960488
char_3_t             91.960488
char_4_t             91.960488
char_5_t             91.960488
char_6_t             91.960488
char_7_t             91.960488
char_8_t             91.960488
char_9_t             91.960488
char_10_t             8.039512
char_1_p              0.000000
group_1               0.000000
char_2_p              0.000000
char_3_p              0.000000
char_4_p              0.000000
char_5_p              0.000000
char_6_p              0.000000
char_7_p              0.000000
char_8_p              0.000000
char_9_p              0.000000
char_10_p             0.000000
char_11               0.000000
char_12               0.000000
char_13               0.000000
char_14               0.000000
char_15               0.000000
char_16               0.000000
char_17               0.000000
char_18               0.000000
char_19               0.000000
char_20 

In [61]:
char_columns_drop = char_columns1.copy()
char_columns_drop.pop()
print(char_columns_drop)

['char_1_t', 'char_2_t', 'char_3_t', 'char_4_t', 'char_5_t', 'char_6_t', 'char_7_t', 'char_8_t', 'char_9_t']


Since the number of NaN is high we shall drop these 9 columns in train and test set

In [62]:
train_df = train_df.drop(char_columns_drop, axis = 1)
test_df = test_df.drop(char_columns_drop, axis = 1)

In [63]:
new_train_df = new_train_df.drop(char_columns_drop, axis = 1)
new_test_df = new_test_df.drop(char_columns_drop, axis = 1)

Now let's impute the values using median for char_10_t. Mean is affected by outliers. So we impute based on median or mode. Let's decide whether we will impute the median for each activity_new or activity_new2 or activity_category or group_1 for the missing values.

In [64]:
temp_df = new_train_df[new_train_df.char_10_t.isna()]

In [65]:
temp_df['activity_new'].unique()

array([1.])

It won't make sense to impute based on median grouped by activity_new. It has only one value: 1.

In [66]:
temp_df['activity_new2'].unique()

array([  9923., 198174., 214090., ..., 245264.,   9232., 330864.])

In [67]:
temp_df['activity_new2'].nunique()

157615

It won't make sense to impute based on median grouped by activity_new2. It has too many values.

In [68]:
temp_df['activity_category'].unique()

array([1.])

It won't make sense to impute based on median grouped by activity_category. It has only one value: 1.

In [69]:
temp_df['group_1'].unique()

array([36096., 17304.,  9439., ..., 27631., 18907.,  2632.])

In [70]:
temp_df['group_1'].nunique()

17008

It will be reasonable to impute based on median grouped by group_1. It has only 17008 values.

In [71]:
temp_df.head()

people_id  activity_category  char_10_t  ...    year_p  month_p  day_p
52   ppl_100025                1.0        NaN  ...      2022        8     26
105  ppl_100033                1.0        NaN  ...      2022        7     26
106  ppl_100033                1.0        NaN  ...      2022        7     26
107  ppl_100033                1.0        NaN  ...      2022        7     26
108  ppl_100033                1.0        NaN  ...      2022        7     26

[5 rows x 51 columns]

In [72]:
grouped = temp_df.groupby(['activity_new'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

activity_new
1.0    41.132506
Name: outcome, dtype: float64

In [73]:
grouped = temp_df.groupby(['activity_new2'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

activity_new2
2.0           0.0
7.0         100.0
12.0          0.0
13.0        100.0
14.0          0.0
16.0          0.0
19.0          0.0
20.0          0.0
21.0          0.0
30.0        100.0
32.0          0.0
36.0          0.0
37.0        100.0
38.0          0.0
40.0          0.0
41.0          0.0
43.0          0.0
46.0        100.0
51.0          0.0
55.0          0.0
57.0          0.0
60.0          0.0
66.0        100.0
71.0        100.0
77.0        100.0
81.0        100.0
92.0        100.0
93.0          0.0
97.0          0.0
100.0         0.0
            ...  
419981.0      0.0
419982.0      0.0
419986.0    100.0
419988.0      0.0
419989.0    100.0
419991.0      0.0
419992.0      0.0
419995.0      0.0
419997.0      0.0
420007.0      0.0
420008.0      0.0
420014.0      0.0
420015.0      0.0
420016.0      0.0
420018.0      0.0
420019.0      0.0
420020.0      0.0
420023.0      0.0
420024.0      0.0
420025.0      0.0
420028.0      0.0
420035.0    100.0
420039.0      0.0
420041.0      

In [74]:
grouped = temp_df.groupby(['year'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

year
2022    45.028289
2023    38.382614
Name: outcome, dtype: float64

In [75]:
grouped = temp_df.groupby(['month'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

month
1     41.322503
2     38.858336
3     37.122771
4     34.878356
5     38.324041
6     41.584826
7     38.141718
8     41.616709
9     55.711862
10    41.368148
11    41.734792
12    42.373892
Name: outcome, dtype: float64

In [76]:
grouped = temp_df.groupby(['day'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

day
1     40.902745
2     43.148956
3     39.182948
4     39.631783
5     38.917596
6     37.500000
7     38.403531
8     37.201633
9     39.162202
10    40.422174
11    32.823246
12    37.226502
13    36.227224
14    51.570955
15    43.794660
16    44.723914
17    39.988772
18    38.028896
19    38.262122
20    43.842365
21    39.255657
22    41.825855
23    41.040219
24    46.090298
25    42.143687
26    41.042909
27    38.874976
28    39.849806
29    44.037294
30    43.341014
31    41.276764
Name: outcome, dtype: float64

In [77]:
grouped = temp_df.groupby(['year_p'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

year_p
2020    47.535820
2021    45.287325
2022    43.934609
2023    33.574583
Name: outcome, dtype: float64

In [78]:
grouped = temp_df.groupby(['month_p'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

month_p
1     38.475059
2     36.073793
3     40.845070
4     38.027580
5     42.045330
6     42.297388
7     41.079323
8     46.543507
9     41.976351
10    42.879891
11    40.093885
12    42.242205
Name: outcome, dtype: float64

In [79]:
grouped = temp_df.groupby(['day_p'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

day_p
1     43.557951
2     42.348163
3     44.189662
4     41.014799
5     43.672457
6     45.208597
7     38.490793
8     41.792528
9     36.613226
10    44.427304
11    30.244296
12    38.016997
13    37.547438
14    40.804416
15    38.438224
16    39.076923
17    36.180157
18    35.780568
19    45.889101
20    44.656085
21    43.496186
22    41.014756
23    40.966808
24    43.804348
25    42.427708
26    43.099822
27    49.720270
28    37.673183
29    41.729446
30    36.620964
31    40.390775
Name: outcome, dtype: float64

In [80]:
grouped = temp_df.groupby(['activity_category'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

activity_category
1.0    41.132506
Name: outcome, dtype: float64

In [81]:
grouped = temp_df.groupby(['char_10_t'])
grouped['outcome'].sum()

Series([], Name: outcome, dtype: int64)

In [82]:
grouped = temp_df.groupby(['char_1_p'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

char_1_p
1    69.335242
2    35.955546
Name: outcome, dtype: float64

In [83]:
grouped = temp_df.groupby(['char_2_p'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

char_2_p
1    69.335242
2     0.000000
3    72.715952
Name: outcome, dtype: float64

In [84]:
grouped = temp_df.groupby(['char_3_p'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

char_3_p
1     67.400177
2     48.543042
3     39.863714
4     45.942879
5     47.324668
6     42.020024
7     54.304482
8     48.016277
9     31.851852
10    43.345488
11    28.368708
12    21.614350
13    37.451737
14    43.673469
15    42.214933
16    38.922156
17    12.603648
18    18.154158
19    32.363162
20    31.990521
21    53.682488
22    61.111111
23    30.985915
24    32.356077
25    20.634921
26    18.974359
27    38.235294
28    32.917706
29    64.224138
30    37.889688
31    13.135593
32    30.147059
33    23.684211
34    38.888889
35    20.833333
36     7.142857
37    25.925926
38     7.692308
39    11.764706
40    32.448284
41    16.666667
44     0.000000
Name: outcome, dtype: float64

In [85]:
grouped = temp_df.groupby(['char_4_p'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

char_4_p
1     67.400177
2     43.900667
3     39.312457
4     46.074074
5     48.084799
6     41.558106
7     43.921354
8     47.916230
9     39.046579
10    53.473399
11    43.706294
12    39.530686
13    50.241546
14    43.518519
15    43.579522
16    25.656325
17    35.582822
18    40.928726
19    39.159292
20    19.565217
21    37.500000
22     6.073753
23    39.336493
24    18.604651
25    32.448284
Name: outcome, dtype: float64

In [86]:
grouped = temp_df.groupby(['char_5_p'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

char_5_p
1    67.400177
2    40.670912
3    43.887370
4    45.959713
5    41.775016
6    38.120202
7    53.791525
8    33.971044
9    32.448284
Name: outcome, dtype: float64

In [87]:
grouped = temp_df.groupby(['char_6_p'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

char_6_p
1    61.128527
2    12.179487
3    41.686215
4    29.034819
5    34.949120
6    40.507860
7     0.000000
Name: outcome, dtype: float64

In [88]:
grouped = temp_df.groupby(['char_7_p'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

char_7_p
1     74.365663
2     63.278512
3     62.235469
4     54.129834
5     18.673705
6      5.437435
7     43.985936
8     36.975155
9     41.353921
10    90.881764
11    52.502765
12    61.619542
13    12.008565
14    29.116430
15    35.048860
16    27.917239
17    31.643707
18    36.561154
19    50.014033
20    41.413996
21    15.169492
22    39.003787
23    86.975242
24    25.194897
25    30.846485
Name: outcome, dtype: float64

In [89]:
grouped = temp_df.groupby(['char_8_p'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

char_8_p
1    37.962771
2    65.735516
3    20.181093
4    57.958889
5    47.085303
6    42.461197
7    51.049292
8    37.263520
Name: outcome, dtype: float64

In [90]:
grouped = temp_df.groupby(['char_9_p'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

char_9_p
1    36.953859
2    65.603322
3    22.177244
4    60.148162
5    48.280255
6    40.399153
7    56.299603
8    36.708861
9    64.497239
Name: outcome, dtype: float64

In [91]:
grouped = temp_df.groupby(['char_10_p'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

char_10_p
0    33.185152
1    67.208083
Name: outcome, dtype: float64

In [92]:
grouped = temp_df.groupby(['char_10_p'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

char_10_p
0    33.185152
1    67.208083
Name: outcome, dtype: float64

In [93]:
grouped = temp_df.groupby(['char_11'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

char_11
0    34.547811
1    66.361156
Name: outcome, dtype: float64

In [94]:
grouped = temp_df.groupby(['char_12'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

char_12
0    34.285479
1    63.921461
Name: outcome, dtype: float64

In [95]:
grouped = temp_df.groupby(['char_13'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

char_13
0    29.281996
1    64.199054
Name: outcome, dtype: float64

In [96]:
grouped = temp_df.groupby(['char_14'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

char_14
0    33.428449
1    64.714521
Name: outcome, dtype: float64

In [97]:
grouped = temp_df.groupby(['char_15'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

char_15
0    32.944932
1    65.595587
Name: outcome, dtype: float64

In [98]:
grouped = temp_df.groupby(['char_16'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

char_16
0    32.426013
1    65.563122
Name: outcome, dtype: float64

In [99]:
grouped = temp_df.groupby(['char_17'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

char_17
0    32.290464
1    65.143800
Name: outcome, dtype: float64

In [100]:
grouped = temp_df.groupby(['char_18'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

char_18
0    35.789924
1    67.293950
Name: outcome, dtype: float64

In [101]:
grouped = temp_df.groupby(['char_19'])
grouped['outcome'].mean()*100

char_19
0    32.067372
1    66.411076
Name: outcome, dtype: float64

In [102]:
grouped = temp_df.groupby(['char_20'])
grouped['outcome'].mean()*100

char_20
0    34.322838
1    66.609512
Name: outcome, dtype: float64

In [103]:
grouped = temp_df.groupby(['group_1'])
grouped['outcome'].sum()/grouped['outcome'].count()*100

group_1
2.0        100.000000
3.0        100.000000
4.0        100.000000
5.0        100.000000
6.0        100.000000
7.0        100.000000
8.0        100.000000
9.0        100.000000
10.0         0.000000
11.0       100.000000
12.0         0.000000
13.0       100.000000
17.0       100.000000
20.0       100.000000
21.0       100.000000
24.0       100.000000
25.0       100.000000
26.0       100.000000
27.0       100.000000
28.0       100.000000
30.0       100.000000
31.0       100.000000
32.0       100.000000
33.0       100.000000
34.0       100.000000
35.0        66.666667
36.0       100.000000
37.0       100.000000
38.0       100.000000
39.0       100.000000
              ...    
51342.0    100.000000
51350.0      0.000000
51352.0      0.000000
51355.0    100.000000
51356.0    100.000000
51362.0      0.000000
51363.0    100.000000
51365.0    100.000000
51369.0    100.000000
51373.0      0.000000
51374.0      0.000000
51379.0      0.000000
51381.0    100.000000
51396.0    100.000000
51

In [104]:
temp_df.head()

people_id  activity_category  char_10_t  ...    year_p  month_p  day_p
52   ppl_100025                1.0        NaN  ...      2022        8     26
105  ppl_100033                1.0        NaN  ...      2022        7     26
106  ppl_100033                1.0        NaN  ...      2022        7     26
107  ppl_100033                1.0        NaN  ...      2022        7     26
108  ppl_100033                1.0        NaN  ...      2022        7     26

[5 rows x 51 columns]

In [105]:
temp_df.outcome.unique()

array([0, 1])

In [106]:
del temp_df

In [107]:
train_df['outcome'].sum()/train_df.shape[0]*100

44.39543965728709

The classes 1 and 0 are not imbalanced. 43.6% of the values are 1s and rest are 0s. Hence, we don't need to do upsampling (stratified or SMOTE) or downsampling

In [108]:
input_features = list(set(train_df.columns.values) & set(test_df.columns.values))

In [109]:
not_useful = ['people_id']
useful = [itm for itm in input_features if itm not in not_useful]

In [110]:
print(input_features)

['char_28', 'char_29', 'activity_new2', 'char_34', 'day', 'char_25', 'char_7_p', 'activity_new', 'year_p', 'char_5_p', 'char_18', 'month', 'char_37', 'char_23', 'char_20', 'char_9_p', 'people_id', 'char_15', 'char_31', 'char_4_p', 'char_26', 'char_1_p', 'char_2_p', 'char_22', 'char_33', 'char_14', 'char_38', 'char_35', 'char_36', 'char_6_p', 'char_13', 'group_1', 'char_16', 'char_17', 'char_11', 'year', 'char_10_t', 'activity_category', 'char_3_p', 'char_12', 'char_32', 'char_8_p', 'char_10_p', 'char_27', 'month_p', 'char_24', 'char_30', 'char_21', 'day_p', 'char_19']


In [111]:
Y_train = train_df['outcome']
X_train = train_df[useful]
X_test = test_df[useful]

In [112]:
not_categorical = ['year', 'month', 'day', 'year_p', 'month_p', 'day_p']
categorical = [itm for itm in X_train.columns.values if itm not in not_categorical]

In [113]:
X_train[categorical].head()

char_28  char_29  activity_new2   ...     char_30  char_21  char_19
0        1        0      1734928.0   ...           1        1        0
1        1        0      2434093.0   ...           1        1        0
2        1        0      3404049.0   ...           1        1        0
3        1        0      3651215.0   ...           1        1        0
4        1        0      4109017.0   ...           1        1        0

[5 rows x 43 columns]

In [114]:
dummies = pd.get_dummies(X_train[categorical], sparse = True)

In [115]:
new_X_train = X_train.drop(categorical, axis = 1)

In [116]:
new_X_train = pd.concat([new_X_train, dummies], axis = 1)

In [117]:
new_X_train.head()

day  year_p  month  year   ...     char_24  char_30  char_21  char_19
0   26    2021      8  2023   ...           0        1        1        0
1   27    2021      9  2022   ...           0        1        1        0
2   27    2021      9  2022   ...           0        1        1        0
3    4    2021      8  2023   ...           0        1        1        0
4   26    2021      8  2023   ...           0        1        1        0

[5 rows x 49 columns]

In [118]:
dummies2 = pd.get_dummies(X_test[categorical], sparse = True)

In [143]:
new_X_test = X_test.drop(categorical, axis = 1)
new_X_test = pd.concat([new_X_test, dummies2], axis = 1)
new_X_test.head()

day  year_p  month  year   ...     char_24  char_30  char_21  char_19
0   20    2022      7  2022   ...           0        1        1        1
1   20    2022      7  2022   ...           0        1        1        1
2   14    2022     10  2022   ...           1        1        1        1
3   27    2022     11  2022   ...           1        1        1        1
4   15    2022     10  2022   ...           1        1        1        1

[5 rows x 49 columns]

1. Now let's develop Logistic regression (with lasso for reducing the number of variables), Random Forest (for reducing the number of variables), and XGBoost. Let's do different combinations of ensemble of the methods to get better results.

In [120]:
from xgboost import XGBClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import roc_auc_score

# split data into train and validation sets
seed = 123
val_size = 0.33
new_X_train2, X_val, Y_train2, Y_val = train_test_split(new_X_train, Y_train, test_size=val_size, random_state=seed)

In [121]:
# fit model no training data
model = XGBClassifier()
model.fit(new_X_train2, Y_train2)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bytree=1, gamma=0, learning_rate=0.1, max_delta_step=0,
       max_depth=3, min_child_weight=1, missing=None, n_estimators=100,
       n_jobs=1, nthread=None, objective='binary:logistic', random_state=0,
       reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
       silent=True, subsample=1)

In [126]:
# make predictions for validation data
y_pred = model.predict(X_val)
predictions = [round(value) for value in y_pred]

In [127]:
# evaluate predictions
roc_auc = roc_auc_score(Y_val, predictions)
print("Area under ROC Curve: %f" % (roc_auc * 100.0))

Area under ROC Curve: 87.463661


Let's check whether 10-fold CV improves the AUC

In [145]:
# make predictions for test data
y_pred_test = model.predict(new_X_test)
test_predictions = [round(value) for value in y_pred_test]

In [146]:
act_test = pd.read_csv('../input/act_test.csv', parse_dates = ['date'])

In [147]:
test_pred_series = pd.Series(v for v in test_predictions)

In [153]:
sub_file = 'submission.csv'
print('Writing submission: ', sub_file)
f = open(sub_file, 'w')
f.write('activity_id,outcome\n')
i = 0
for id in act_test['activity_id']:
    str1 = str(id) + ',' + str(test_predictions[i])
    str1 += '\n'
    i += 1
    f.write(str1)
f.close()

Writing submission:  submission.csv


from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import cross_val_score

# 10-fold CV model
kfold = StratifiedKFold(n_splits=10, random_state=123)
results = cross_val_score(model, new_X_train, Y_train, cv=kfold)

print("Area under ROC Curve: %f" % (results.roc_auc * 100.0))

from sklearn.linear_model import LogisticRegression

clf = LogisticRegression(penalty="l1")
clf.fit(new_X_train,Y_train)